### PR TITLE
[BAHIR-107] Upgrade to Scala 2.12 and Spark 2.4.0

### PR DESCRIPTION
**Pull request by lukasz-antoniak**
JIRA ticket: https://issues.apache.org/jira/browse/BAHIR-107.

### CODE REVIEW

**Comment by lresende** [comment]
Could you start a conversation on the dev list regarding the direction this is going to take, are we dropping 2.10 and continue support with 211 and 212 but making 212 the default, and all other details... at least to be informative to the dev community

**Comment by lukasz-antoniak** [comment]
Rebased with latest changes form master branch. I have also fixed `skipTests` handling in the release script (it was not working before) and added another stage to actually execute unit tests with Scala 2.11 and 2.12 before publishing. Can be easily removed if you disagree.

File: dev/release-build.sh [hunk 1]
```diff
@@ -231,83 +229,92 @@ function checkout_code {
...
-    cd .. #exit bahir
+    cd .. # exit bahir
 
     if [ -z "$DRY_RUN" ]; then
         cd "$BASE_DIR/target/bahir"
         git checkout $RELEASE_TAG
         git clean -d -f -x
 
-        $MVN $PUBLISH_PROFILES clean install -DskiptTests -Darguments="-DskipTests"
+        $MVN $PUBLISH_PROFILES clean install -DskipTests=true
```
Comment (lresende) [inline dev/release-build.sh:260]: My understanding is that -DskiptTests does not require a value as described in [Maven docs](http://maven.apache.org/surefire/maven-surefire-plugin/examples/skipping-tests.html).

Also, -arguments is required to pass arguments to internal plugins such as the release plugin.

**Review by lresende — CHANGES_REQUESTED** [review]

File: dev/release-build.sh (hunk 1 above)
Comment (lukasz-antoniak) [reply dev/release-build.sh:260]: `-DskipTests` does not require arguments, but Scala tests do not honour it, see: https://github.com/scalatest/scalatest/issues/466. On my machine, while running commands from original version of release script, tests are were being executed.

**Review by lresende — APPROVED** [review]

**Comment by steventay** [comment]
Hi,

Apologies for commenting here but my emails are getting bounced from the user mailing list.

In any case I'm trying to reference the Spark Streaming lib and am targeting Spark 2.4 / Scala 2.12. From the 2.4.0-SNAPSHOT  documentation, the maven dependency is declared as: 

```
<dependency>
    <groupId>org.apache.bahir</groupId>
    <artifactId>spark-sql-streaming-akka_2.11</artifactId>
    <version>2.4.0-SNAPSHOT</version>
</dependency>
```

It seems that we are still referencing 2.11. Shouldn't that be 2.12 instead of 2.11? 

Also looking around maven central and Bahir’s download page http://bahir.apache.org/downloads/spark/, I can’t see 2.4.0-SNAPSHOT being hosted anywhere. 
Is there anywhere the 2,4.0 lib can be downloaded from? I guess I can build from source but just trying to avoid that if a binary version is available for download. 

Thanks,
Steven

**Comment by jacek-rzrz** [comment]
I cannot find the Scala 2.12 / Spark 2.4 version of spark-streaming-pubsub in Maven central yet. Is there a chance it will be released soon?

**Comment by mahtuog** [comment]
Any info on when the sql-streaming-mqtt for Spark 2.4 will be released ?

**Comment by WinterSoldier13** [comment]
Sorry to bother you guys, but does `bahir2.4.0` supports `spark2.4.0`? On the website https://bahir.apache.org/ the version of Bahir shows as `2.3.4` but in maven `2.4.0` is available, though I am getting error messages while trying to run it with `spark 2.4.0`.
Kindly see: https://stackoverflow.com/questions/65252433/does-bahir2-4-0-supports-spark2-4-0 
Thanks.

**Comment by lresende** [comment]
I have just published a snapshot of Bahir 2.4.0 and will work on a release for the weekend.

**Pull request closed without merging**